dev-infra/release/env-stamp.ts

```
@@ -0,0 +1,64 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {execSync as _execSync} from 'child_process';
+
+/**
+ * Log the environment variables expected by bazel for stamping.
+ *
+ * See the section on stamping in docs / BAZEL.md
+ *
+ * This script must be a NodeJS script in order to be cross-platform.
+ * See https://github.com/bazelbuild/bazel/issues/5958
+ * Note: git operations, especially git status, take a long time inside mounted docker volumes
+ * in Windows or OSX hosts (https://github.com/docker/for-win/issues/188).
+ */
+export function buildEnvStamp() {
+  console.info(`BUILD_SCM_BRANCH ${getCurrentBranch()}`);
+  console.info(`BUILD_SCM_COMMIT_SHA ${getCurrentSha()}`);
+  console.info(`BUILD_SCM_HASH ${getCurrentSha()}`);
+  console.info(`BUILD_SCM_LOCAL_CHANGES ${hasLocalChanges()}`);
+  console.info(`BUILD_SCM_USER ${getCurrentGitUser()}`);
+  console.info(`BUILD_SCM_VERSION ${getSCMVersion()}`);
+  process.exit(0);
+}
+
+/** Run the execSync command and return the stdout as a trimmed string. */
+function execSync(cmd: string) {
+  return _execSync(cmd).toString().trim();
+}
+
+/** Whether the repo has local changes. */
+function hasLocalChanges() {
+  return !execSync(`git status --untracked-files=no --porcelain`);
+}
+
+/** Get the version based on the most recent semver tag. */
+function getSCMVersion() {
+  const version = execSync(`git describe --match [0-9]*.[0-9]*.[0-9]* --abbrev=7 --tags HEAD`);
```

In the components repo we just rely on the version from the `package.json`. Do you know why we need to do find it through Git? I know that framework did this in the past, but I'm wondering if it's time to consider the other more predictable approach.

---

I believe this is done within `angular/angular` still to allow for the version to be checked to verify that the build being created with a stamp, is actually the tagged release.  If the current repository state is not the exact sha that is the tag point, then it does not have the version of the sha.

Additionally, this allows for us to have stamping within our build artifacts created during a run on CI without it being tagged in our `package.json`s that are created as the actually released version.


Its definitely something we should step back and consider for what is our best option.  I think this is something that definitely makes sense to unify on, to allow for better tooling both via `ng-dev` as well as our `bazel` macros/rules

---

That makes sense. The thing I'm slightly worried about is that the stamping relies on Git to determine the _current_ version. With the `package.json` it is super easy and predictable to tell what the release will be. That also matches what's usually the process when building and publishing through NPM. i.e. Updating the version in the `package.json`, building and publishing.

It's totally valid to create the tag after the release packages have been built (build could theoretically fail; and then a tag is not desired). 

The benefit for CI snapshot builds that you mentioned is definitely a good one. We could chat about this as a follow-up. I just want to make clear that components repo currently uses a different process and that I don't know whether this is too unpredictable through Git (see how it silently used the wrong version when we started for v10.x.x)

---

How do you feel about landing this as is for now, and then we can determine what flow/option is best to standardize on long term in follow up work?

---

Yeah, 100%. I also proposed above that we can chat about this in a follow-up 😄